Tests to increase test/code coverage for MetaDict

m-charlton opened the pull request:
I'm looking to learn more about the sunpy codebase and thought a good idea 
would be to start with the unit tests. To this end I've added some tests
for `sunpy.util.metadata.MetaDict`.

There are a lot of tests but, I've aimed for test coverage as well as code
coverage.

I'll start with `sunpy.util` and submit more than one file at a time but thought 
that it would be useful to get some feedback on this initial commit before I go
any further. 

This is my first use of `pytest` so I may not be using the correct idioms.

Michael.

pep8speaks commented:
Hello @m-charlton! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`sunpy/util/tests/test_metadata.py`](https://github.com/m-charlton/sunpy/blob/cba5cc4694a2c35dd634362e990da0113de64fcf/sunpy/util/tests/test_metadata.py) !

Cadair commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,205 @@
+
+    def setup_method(self):
+
+        self.seas = [['labrador', 'americas'],
```
Generally when using pytest instead of writing classes, you write functions and use pytest fixtures to provide input to the tests. http://doc.pytest.org/en/latest/fixture.html

m-charlton replied:
Cheers, will make changes to use fixtutres

m-charlton commented:
I've changed to use `pytest`. This is my first attempt at using `pytest` so feedback would 
be appreciated.

Tested with python2.7 & python3.5.

Cadair commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,225 @@
+    return MetaDict(std_test_data)
+
+
+class TestInit:
```
There is no need for these to be in classes when using pytest either. You should keep the docstrings as delimiters and comments, but pytest will collect all functions starting with `test_`.

m-charlton replied:
Cheers, will have another refactor

Cadair commented:
Thanks for the tests @m-charlton !

m-charlton commented:
Removed classes

Cadair commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,224 @@
+    actual against expected contents and ensure insertion order is
+    preserved.
+
+    :param metadict_inst: a MetaDict
```
SunPy uses the numpydoc style for it's docstrings, so if you could change this to look follow these rules: http://docs.astropy.org/en/stable/development/docrules.html#sections

Cadair replied:
Which also includes a line break after the `"""`

m-charlton replied:
Apologies for the delay - will update to conform.

Cadair commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,224 @@
+def test_has_key_method(std_metadict):
+
+    # suppress pep8/flake8 'W601 .has_key() is deprecated, ...' warnings
+    # as MetaDict explicitly supports the 'has_key()' method
```
This is interesting. Do we want to deprecate `has_key` on our `MetaDict` as well? @Alex-Ian-Hamilton ??

wafels replied:
This is a separate issue.

ehsteve commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,247 @@
+    assert normalise_keys(metadict_inst.items()) == normalise_keys(expected)
+
+
+def check(metadict_inst, expected):
```
This function name needs to be more descriptive.

m-charlton replied:
Will restructure tests cases and provide comments better describing 
test intent.

I'll also rename, and document, supporting test functions.

dpshelio commented on sunpy/util/tests/test_metadata.py:
```diff
@@ -0,0 +1,247 @@
+def test_setdefault_method(std_metadict):
+    std_metadict.setdefault('poseidon', 'NOT-FOUND')
+    assert std_metadict.get('Poseidon') == 'NOT-FOUND'
+    assert std_metadict['pOSeidon'] == 'NOT-FOUND'
```
I would suggest having a test that only checks cases of the keys, and the tests above just checking what their name refer to.

m-charlton replied:
Will revisit.

m-charlton commented:
### Changes following code review  #1984

As per @ehsteve

#### Structure of test code re-organized:

- Supporting function `check` renamed to more descriptive `check_contents_and_insertion_order`.

- Docstrings of supporting functions added/expanded.

- Split the test data in to two different classes - `sea_locations` & `atomic_weights`.
  This will hopefully make it less difficult to reason about what happens when 
  different `MetaDict` instances are merged together.

- Fixtures, `seas_and_atomic_weights` & `combined_seas_atomic` moved closer 
  to their point of use.

##### Test cases organised as follows:

Different ways of initialising a `MetaDict`:

```
test_init_with_lst_of_lsts
test_init_with_tuple_of_tuples
test_init_with_dict
test_init_with_metadict
test_init_with_illegal_arg
```

Individual methods 

```
test_membership
test_getitem
test_setitem_existing_entry
test_setitem_new_entry
test_get
test_setdefault
test_has_key
test_pop

test_update_with_distinct_keys
test_update_with_like_keys
test_update_with_dict
test_update_with_same_metadict
```
   
Following suggestion from @dsphelio, moved testing of case insensitivity out to 

```
test_key_case_insensitivity
```

dpshelio commented:
@ehsteve  @Cadair  - what's the problem with this?  it looks good to me.
